Add horizontal error bars and linear regression

paul019 opened the pull request:
- change parser including parser logic (see README)
- update README accordingly
- update sample data accordingly
- update Measurement class to hold errors for x- and y-value
- update drawer to draw horizontal error bars

paul019 committed:
Add horizontal error bars

- change parser including parser logic (see README)
- update README accordingly
- update sample data accordingly
- update Measurement class to hold errors for x- and y-value
- update drawer to draw horizontal error bars

paul019 committed:
Prepare linear regression

paul019 committed:
Draw linear regression

paul019 committed:
Change regression stroke width

paul019 committed:
Modularize the code even more to be able to implement error lines more easily

paul019 committed:
Add capability to draw error lines

paul019 committed:
Update README and change config parameter names

paul019 committed:
Improve console output

Splines commented on README.md:
```diff
@@ -37,13 +37,18 @@ File paths are *relative* (see also `Sample usage` above).
+| 6 | upper error of y-value |
 
-\*Note: If only three columns are supplied, the third column's content is interpreted as a *symmetrical* error of the y-value.
+If don't want your x- or y-values to have error bars, simply set the error to `0.0`. However, you still need to fill the corresponding column.
```
If *you*...

Splines replied:
And the corresponding column**s**

paul019 replied:
Done

Splines commented on data/data.csv:
```diff
@@ -1,3 +1,4 @@
-200,64.7,0.5,1.5
-250,80.5,0.5,1.5
\ No newline at end of file
+0.481,0.002,0.002,200,2,2
```
I think at this point it might be worth to actually include a CSV header and just skip the first line during parsing. Otherwise, it's tedious to always look into the documentation when dealing with the data files.

paul019 replied:
Done

Splines commented on src/drawer.py:
```diff
@@ -15,7 +19,9 @@ class Drawer:
 
-    def __init__(self, trafo: Transformer, config) -> None:
+    def __init__(self, trafo: Transformer, config, regressionConfig) -> None:
+        self.regressionConfig = regressionConfig
```
snake case, see [here](https://peps.python.org/pep-0008/#function-and-variable-names) and probably more useful [here](https://stackoverflow.com/a/8423697/9655481)

paul019 replied:
Done

Splines commented on src/drawer.py:
```diff
@@ -163,3 +205,50 @@ def _draw_axis(self, axis: Axis):
+
+        print()
+        print('Linear regression:')
+        print('y = m * x + n')
+        print('m  = {}'.format(m))
+        print('Δm = {}'.format(dm))
+        print('n  = {}'.format(n))
+        print('Δn = {}'.format(dn))
```
where do I find this in the PAP script again?

paul019 replied:
Page 10

Splines commented on src/drawer.py:
```diff
@@ -163,3 +205,50 @@ def _draw_axis(self, axis: Axis):
+    
+    def _draw_linear_function(self, f: LinearFunction, color: tuple[float, float, float]):
+        """
+        Draw linear function of the form y = m * x + b.
```
Draw**s** **a**

paul019 replied:
Done

Splines commented on src/measurement.py:
```diff
@@ -7,26 +18,22 @@ class Measurement:
 
-    def has_error_bounds(self) -> bool:
-        return self.lower_error != 0.0 or self.upper_error != 0.0
+    #def has_error_bounds(self) -> bool:
+    #    return self.lower_error != 0.0 or self.upper_error != 0.0
```
Is this still needed? If not, get rid of it. If yes, uncomment it.

paul019 replied:
Done

Splines commented on src/measurement.py:
```diff
@@ -7,26 +18,22 @@ class Measurement:
-        return f'Point ({self.x}, {self.y}) {errors_str}'
+        #errors_str = f' Errors: ({self.lower_error}, {self.upper_error})'\
+        #    if self.has_error_bounds() else ''
+        return f'Point ({self.x}, {self.y})'
```
still needed?

paul019 replied:
My bad. An artifact of bug fixing.

Splines commented on src/linear_regressor.py:
```diff
@@ -0,0 +1,79 @@
+import math
+from src.measurement import Measurement
+
+class LinearRegression:
```
That's a really nice encapsulation here.

paul019 replied:
Thanks :-)

Splines commented on src/transformer.py:
```diff
@@ -175,11 +178,23 @@ def get_pdf_coords_for_point_on_axis(self, axis: Axis, value: float) -> tuple[fl
+            label += self.points_offset_y
+            return label
+    
+    def get_linear_regression(self):
+        return do_linear_regression(self.measurements)
```
Not too sure if transformer should be worried with the linear regression. Based on what scope we have right now for the transformer

> Transforms measurements into PDF coordinates based on grid configuration. Note that analyze_and_offset_measurements must be called first.

I'd say that a linear regression does not fit in this scope. Maybe construct a linear regression directly in `drawer.py` in line `48`? Not sure if this is a good idea though.

paul019 replied:
I find it quite fitting for the transformer. Then the drawer is only worried with drawing and the transformer is there for any data processing.

Splines commented on src/measurement.py:
```diff
@@ -7,26 +18,22 @@ class Measurement:
     """
 
-    def __init__(self, x: float, y: float):
+    def __init__(self, x: float, x_lower_error: float, x_upper_error: float, y: float, y_lower_error: float, y_upper_error: float):
```
line too long

paul019 replied:
Fixed

Splines commented on src/measurement.py:
```diff
@@ -7,26 +18,22 @@ class Measurement:
     and are not drawn.
     """
```
A measurement `may` define error bounds. Now it has to define error bounds, right?

Splines replied:
And they don't default to `0.0` anymore.

paul019 replied:
Fixed

Splines commented on src/measurement.py:
```diff
@@ -1,3 +1,14 @@
+class Error:
```
Docstring missing

paul019 replied:
Fixed

Splines commented on src/transformer.py:
```diff
@@ -175,11 +178,23 @@ def get_pdf_coords_for_point_on_axis(self, axis: Axis, value: float) -> tuple[fl
-
-        return (grid_coord-offset)/scale + point_offset
+        if axis == Axis.HORIZONTAL:
+            label = self.get_data_point_from_grid_coords(grid_coord, 0)[0]
+            label += self.points_offset_x
+            return label
```
Why not directly return? (not too relevant though as it looks fine)

Splines reviewed (changes requested):
That's a nice new feature, thanks for implementing this @paul019. Here are my two cents ;)

Splines commented on src/measurement.py:
```diff
@@ -1,3 +1,14 @@
+    
+    def avg_error(self):
+        return (self.lower_error + self.upper_error) / 2.0
+
```
To be really nitpicky: my linter tells me to add two empty lines before a new class starts.

Splines commented on src/measurement.py:
```diff
@@ -1,3 +1,14 @@
+
+    def exists(self):
+        return self.lower_error != 0.0 or self.upper_error != 0.0
+    
```
To be really nitpicky: my linter tells me to remove the unnecessary `tab` character in this line.

Splines reviewed (changes requested):
some more nitpicky stuff

Splines commented on src/linear_regressor.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    Makes a linear regression.
+    """
+    x = [m.x for m in measurements]
+    y = [m.y for m in measurements]
```
This is the second time we need this (see other occurrence in `analyze_and_offset_measurements()`), so it might be a good idea to put this as method into the measurement itself, e.g. `x,y = m.get_data()` or something like that.

Splines commented:
One thing to mention: we should probably somehow note (e.g. in the release notes) that this is a breaking change since the `CSV` format changed and is not compatible with the previous format anymore. In a production-like environments, one would even consider writing a small script that helps converting old file formats to not alienate users, but that's a little overkill for this project I guess.

paul019 committed:
Fix a bug

Formerly, the regressor used the offset datapoints thus resulting in a wrong value for "n"

paul019 committed:
Use a linter

paul019 committed:
Fix typos in README

paul019 committed:
Automatically detect whether the csv file has a header row

paul019 committed:
Fix missing snake case

paul019 committed:
Fix typo in comment

paul019 committed:
Add missing docstring

paul019 committed:
Fix error in comment

paul019 committed:
Merge branch 'main' into feauture/horizontal-error-bars

paul019 committed:
Update gitignore

paul019 committed:
Add better testing data

paul019 committed:
Merge branch 'feauture/horizontal-error-bars' of https://github.com/paul019/pappe into feauture/horizontal-error-bars

paul019 committed:
Only do regression when it is needed

paul019 committed:
Fix a severe bug in the transformer

Formerly, data and offset data was interchanged at some point. This bug fix also makes the language used clearer.

paul019 committed:
Add feature to hide outliers from the linear regressor
